src/Symfony/Component/Dotenv/Dotenv.php

```diff
@@ -350,7 +352,7 @@ private function resolveVariables($value)
         }
 
         $regex = '/
-            (\\\\)?                    # escaped with a backslash?
+            ((?:\\\\)?\\\\)?           # escaped with a backslash?
```

Comment:
Thinking more about it this will probably not be enough as there could be an endless number of backslashes and we need to check whether we have an even or an odd number of them to decide whether or not the `$` character is escaped.